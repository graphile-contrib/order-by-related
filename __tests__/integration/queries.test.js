// @ts-check
const { printSchema } = require("postgraphile/graphql");
const { withPgClient, makePreset } = require("../helpers.js");
const { makeSchema } = require("postgraphile");
const { grafast, isAsyncIterable } = require("postgraphile/grafast");
const { readdirSync } = require("fs");
const { readFile } = require("fs/promises");
const { resolve: resolvePath } = require("path");
const debug = require("debug")("graphile-build:schema");

const queriesDir = `${__dirname}/../fixtures/queries`;
const queryFileNames = readdirSync(queriesDir);
let queryResults = [];

const kitchenSinkData = () => readFile(`${__dirname}/../p-data.sql`, "utf8");

beforeAll(() => {
  // Get a few GraphQL schema instance that we can query.
  const gqlSchemasPromise = (async () => {
    // Different fixtures need different schemas with different configurations.
    // Make all of the different schemas with different configurations that we
    // need and wait for them to be created in parallel.
    const [normal, columnAggregates] = await Promise.all([
      makeSchema(makePreset(["p"], {})),
      makeSchema(
        makePreset(["p"], {
          graphileBuildOptions: {
            orderByRelatedColumnAggregates: true,
          },
        })
      ),
    ]);
    debug(printSchema(normal.schema));
    return {
      normal,
      columnAggregates,
    };
  })();

  // Execute all of the queries in parallel. We will not wait for them to
  // resolve or reject. The tests will do that.
  //
  // All of our queries share a single client instance.
  const queryResultsPromise = (async () => {
    // Wait for the schema to resolve. We need the schema to be introspected
    // before we can do anything else!
    const gqlSchemas = await gqlSchemasPromise;
    // Get a new Postgres client instance.
    return await withPgClient(async (pgClient) => {
      // Add data to the client instance we are using.
      await pgClient.query(await kitchenSinkData());
      // Run all of our queries in parallel.
      return await Promise.all(
        queryFileNames.map(async (fileName) => {
          // Read the query from the file system.
          const source = await readFile(
            resolvePath(queriesDir, fileName),
            "utf8"
          );
          // Get the appropriate GraphQL schema for this fixture. We want to test
          // some specific fixtures against a schema configured slightly
          // differently.
          const schemas = {
            "columnAggregates.graphql": gqlSchemas.columnAggregates,
          };
          const { schema, resolvedPreset } =
            schemas[fileName] ?? gqlSchemas.normal;
          // Return the result of our GraphQL query.
          const result = await grafast({
            schema,
            source,
            resolvedPreset,
            requestContext: {},
          });
          if (isAsyncIterable(result)) {
            throw new Error(`Didn't expect an async iterable`);
          }
          if (result.errors) {
            console.log(result.errors.map((e) => e.originalError ?? e));
          }
          return result;
        })
      );
    });
  })();

  // Flatten out the query results promise.
  queryResults = queryFileNames.map(
    async (_, i) => (await queryResultsPromise)[i]
  );
});

for (let i = 0; i < queryFileNames.length; i++) {
  // eslint-disable-next-line jest/valid-title
  test(queryFileNames[i], async () => {
    expect(await queryResults[i]).toMatchSnapshot();
  });
}
