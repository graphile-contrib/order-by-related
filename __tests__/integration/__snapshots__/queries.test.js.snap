// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`order-by-related.graphql 1`] = `
Object {
  "data": Object {
    "A": Object {
      "nodes": Array [
        Object {
          "id": 1,
        },
        Object {
          "id": 2,
        },
        Object {
          "id": 3,
        },
        Object {
          "id": 4,
        },
        Object {
          "id": 5,
        },
      ],
    },
    "BACKWARD_BY_FOO_ID__NAME_ASC": Object {
      "nodes": Array [
        Object {
          "backwardByFooId": Object {
            "name": "backward1",
          },
          "id": 1,
        },
        Object {
          "backwardByFooId": Object {
            "name": "backward2",
          },
          "id": 2,
        },
        Object {
          "backwardByFooId": Object {
            "name": "backward3",
          },
          "id": 3,
        },
        Object {
          "backwardByFooId": Object {
            "name": "backward4",
          },
          "id": 4,
        },
        Object {
          "backwardByFooId": null,
          "id": 5,
        },
      ],
    },
    "BACKWARD_BY_FOO_ID__NAME_DESC": Object {
      "nodes": Array [
        Object {
          "backwardByFooId": null,
          "id": 5,
        },
        Object {
          "backwardByFooId": Object {
            "name": "backward4",
          },
          "id": 4,
        },
        Object {
          "backwardByFooId": Object {
            "name": "backward3",
          },
          "id": 3,
        },
        Object {
          "backwardByFooId": Object {
            "name": "backward2",
          },
          "id": 2,
        },
        Object {
          "backwardByFooId": Object {
            "name": "backward1",
          },
          "id": 1,
        },
      ],
    },
    "BACKWARD_COMPOUND_BY_BACKWARD_COMPOUND_1_AND_BACKWARD_COMPOUND_2__NAME_ASC": Object {
      "nodes": Array [
        Object {
          "backwardCompoundByBackwardCompound1AndBackwardCompound2": Object {
            "name": "backwardCompound11",
          },
          "id": 1,
        },
        Object {
          "backwardCompoundByBackwardCompound1AndBackwardCompound2": Object {
            "name": "backwardCompound12",
          },
          "id": 2,
        },
        Object {
          "backwardCompoundByBackwardCompound1AndBackwardCompound2": Object {
            "name": "backwardCompound21",
          },
          "id": 3,
        },
        Object {
          "backwardCompoundByBackwardCompound1AndBackwardCompound2": Object {
            "name": "backwardCompound22",
          },
          "id": 4,
        },
        Object {
          "backwardCompoundByBackwardCompound1AndBackwardCompound2": null,
          "id": 5,
        },
      ],
    },
    "BACKWARD_COMPOUND_BY_BACKWARD_COMPOUND_1_AND_BACKWARD_COMPOUND_2__NAME_DESC": Object {
      "nodes": Array [
        Object {
          "backwardCompoundByBackwardCompound1AndBackwardCompound2": null,
          "id": 5,
        },
        Object {
          "backwardCompoundByBackwardCompound1AndBackwardCompound2": Object {
            "name": "backwardCompound22",
          },
          "id": 4,
        },
        Object {
          "backwardCompoundByBackwardCompound1AndBackwardCompound2": Object {
            "name": "backwardCompound21",
          },
          "id": 3,
        },
        Object {
          "backwardCompoundByBackwardCompound1AndBackwardCompound2": Object {
            "name": "backwardCompound12",
          },
          "id": 2,
        },
        Object {
          "backwardCompoundByBackwardCompound1AndBackwardCompound2": Object {
            "name": "backwardCompound11",
          },
          "id": 1,
        },
      ],
    },
    "CHILDREN_BY_FOO_ID__COUNT_ASC": Object {
      "nodes": Array [
        Object {
          "childrenByFooId": Object {
            "totalCount": 0,
          },
          "id": 3,
        },
        Object {
          "childrenByFooId": Object {
            "totalCount": 0,
          },
          "id": 4,
        },
        Object {
          "childrenByFooId": Object {
            "totalCount": 0,
          },
          "id": 5,
        },
        Object {
          "childrenByFooId": Object {
            "totalCount": 2,
          },
          "id": 1,
        },
        Object {
          "childrenByFooId": Object {
            "totalCount": 2,
          },
          "id": 2,
        },
      ],
    },
    "CHILDREN_BY_FOO_ID__COUNT_DESC": Object {
      "nodes": Array [
        Object {
          "childrenByFooId": Object {
            "totalCount": 2,
          },
          "id": 1,
        },
        Object {
          "childrenByFooId": Object {
            "totalCount": 2,
          },
          "id": 2,
        },
        Object {
          "childrenByFooId": Object {
            "totalCount": 0,
          },
          "id": 3,
        },
        Object {
          "childrenByFooId": Object {
            "totalCount": 0,
          },
          "id": 4,
        },
        Object {
          "childrenByFooId": Object {
            "totalCount": 0,
          },
          "id": 5,
        },
      ],
    },
    "FORWARD_BY_FORWARD_ID__NAME_ASC": Object {
      "nodes": Array [
        Object {
          "forwardByForwardId": Object {
            "name": "forward1",
          },
          "id": 1,
        },
        Object {
          "forwardByForwardId": Object {
            "name": "forward2",
          },
          "id": 2,
        },
        Object {
          "forwardByForwardId": Object {
            "name": "forward3",
          },
          "id": 3,
        },
        Object {
          "forwardByForwardId": Object {
            "name": "forward4",
          },
          "id": 4,
        },
        Object {
          "forwardByForwardId": null,
          "id": 5,
        },
      ],
    },
    "FORWARD_BY_FORWARD_ID__NAME_DESC": Object {
      "nodes": Array [
        Object {
          "forwardByForwardId": null,
          "id": 5,
        },
        Object {
          "forwardByForwardId": Object {
            "name": "forward4",
          },
          "id": 4,
        },
        Object {
          "forwardByForwardId": Object {
            "name": "forward3",
          },
          "id": 3,
        },
        Object {
          "forwardByForwardId": Object {
            "name": "forward2",
          },
          "id": 2,
        },
        Object {
          "forwardByForwardId": Object {
            "name": "forward1",
          },
          "id": 1,
        },
      ],
    },
    "FORWARD_COMPOUND_BY_FORWARD_COMPOUND_1_AND_FORWARD_COMPOUND_2__NAME_ASC": Object {
      "nodes": Array [
        Object {
          "forwardCompoundByForwardCompound1AndForwardCompound2": Object {
            "name": "forwardCompound11",
          },
          "id": 1,
        },
        Object {
          "forwardCompoundByForwardCompound1AndForwardCompound2": Object {
            "name": "forwardCompound12",
          },
          "id": 2,
        },
        Object {
          "forwardCompoundByForwardCompound1AndForwardCompound2": Object {
            "name": "forwardCompound21",
          },
          "id": 3,
        },
        Object {
          "forwardCompoundByForwardCompound1AndForwardCompound2": Object {
            "name": "forwardCompound22",
          },
          "id": 4,
        },
        Object {
          "forwardCompoundByForwardCompound1AndForwardCompound2": null,
          "id": 5,
        },
      ],
    },
    "FORWARD_COMPOUND_BY_FORWARD_COMPOUND_1_AND_FORWARD_COMPOUND_2__NAME_DESC": Object {
      "nodes": Array [
        Object {
          "forwardCompoundByForwardCompound1AndForwardCompound2": null,
          "id": 5,
        },
        Object {
          "forwardCompoundByForwardCompound1AndForwardCompound2": Object {
            "name": "forwardCompound22",
          },
          "id": 4,
        },
        Object {
          "forwardCompoundByForwardCompound1AndForwardCompound2": Object {
            "name": "forwardCompound21",
          },
          "id": 3,
        },
        Object {
          "forwardCompoundByForwardCompound1AndForwardCompound2": Object {
            "name": "forwardCompound12",
          },
          "id": 2,
        },
        Object {
          "forwardCompoundByForwardCompound1AndForwardCompound2": Object {
            "name": "forwardCompound11",
          },
          "id": 1,
        },
      ],
    },
    "PARENT_BY_PARENT_ID__NAME_ASC": Object {
      "nodes": Array [
        Object {
          "id": 1,
          "parentByParentId": Object {
            "name": "parent1",
          },
        },
        Object {
          "id": 2,
          "parentByParentId": Object {
            "name": "parent1",
          },
        },
        Object {
          "id": 3,
          "parentByParentId": Object {
            "name": "parent2",
          },
        },
        Object {
          "id": 4,
          "parentByParentId": Object {
            "name": "parent2",
          },
        },
        Object {
          "id": 5,
          "parentByParentId": null,
        },
      ],
    },
    "PARENT_BY_PARENT_ID__NAME_DESC": Object {
      "nodes": Array [
        Object {
          "id": 5,
          "parentByParentId": null,
        },
        Object {
          "id": 3,
          "parentByParentId": Object {
            "name": "parent2",
          },
        },
        Object {
          "id": 4,
          "parentByParentId": Object {
            "name": "parent2",
          },
        },
        Object {
          "id": 1,
          "parentByParentId": Object {
            "name": "parent1",
          },
        },
        Object {
          "id": 2,
          "parentByParentId": Object {
            "name": "parent1",
          },
        },
      ],
    },
  },
}
`;
